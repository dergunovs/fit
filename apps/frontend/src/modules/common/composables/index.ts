import { ref, computed, onMounted } from 'vue';
import { useRouter, useRoute } from 'vue-router';

import LayoutAdmin from '@/common/components/LayoutAdmin.vue';
import LayoutDefault from '@/common/components/LayoutDefault.vue';

export function useLayout() {
  const router = useRouter();
  const route = useRoute();

  const isLoaded = ref(false);

  const layoutComponent = computed(() => (route.meta.layout === 'admin' ? LayoutAdmin : LayoutDefault));

  onMounted(async () => {
    await router.isReady();
    isLoaded.value = true;
  });

  return {
    isLoaded,
    layoutComponent,
  };
}

export function useRouteId(title: string, isQuery?: boolean) {
  const route = useRoute();

  const id = computed(() => {
    const value = isQuery ? route.query[title]?.toString() : route.params[title].toString();

    return value || '';
  });

  return { id };
}
