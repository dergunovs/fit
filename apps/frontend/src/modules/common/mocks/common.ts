import { vi } from 'vitest';

export const spyCopyToClipboard = vi.spyOn(navigator.clipboard, 'writeText').mockResolvedValue();
