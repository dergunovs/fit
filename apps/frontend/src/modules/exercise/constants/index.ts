export const URL_EXERCISE = '/exercises';
export const URL_EXERCISE_CREATE = `${URL_EXERCISE}/create`;
export const URL_EXERCISE_EDIT = `${URL_EXERCISE}/edit`;

export const EXERCISE_REPEATS_OPTIONS = [8, 10, 12, 15, 20];
export const EXERCISE_REPEATS_DEFAULT = 12;

export const EXERCISE_STATISTICS_HEADERS = [{ title: 'Упражнение' }, { title: 'Подходы' }, { title: 'Повторы' }];
export const MUSCLE_GROUP_STATISTICS_HEADERS = [{ title: 'Мышцы' }, { title: 'Подходы' }, { title: 'Повторы' }];
export const EXERCISE_LIST_HEADERS = [{ title: 'Название' }];

export const EXERCISE_DELETED_TITLE = 'Упражнение удалено';
