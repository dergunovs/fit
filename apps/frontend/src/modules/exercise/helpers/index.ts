import {
  IExercise,
  IExerciseChoosen,
  IExerciseDone,
  IExerciseStatistics,
  IUser,
  IUserEquipment,
} from 'fitness-tracker-contracts';
import { createTempId } from 'mhz-helpers';

import { ITimelineStep } from '@/activity/interface';
import { getWeightsForUserEquipment } from '@/equipment/helpers';

export function getUserEquipmentParams(exercise: IExercise, user?: IUser) {
  const isExerciseHasEquipment = !!exercise.equipment;
  const isExerciseHasEquipmentForWeight = !!exercise.equipmentForWeight?.length;
  const isWeightsRequired = !!exercise.isWeightsRequired;

  const isUserHasEquipment = !!user?.equipments?.some(
    (equipment) => equipment.equipment?._id?.toString() === exercise.equipment?._id?.toString()
  );

  const isUserHasEquipmentForWeight = !!user?.equipments?.some((equipment) =>
    exercise.equipmentForWeight?.some(
      (equipmentForWeight) => equipmentForWeight._id?.toString() === equipment.equipment?._id?.toString()
    )
  );

  return {
    isExerciseHasEquipment,
    isExerciseHasEquipmentForWeight,
    isWeightsRequired,
    isUserHasEquipment,
    isUserHasEquipmentForWeight,
  };
}

export function isUserEquipmentMatches(exercise: IExercise, user?: IUser) {
  let result = false;

  const {
    isExerciseHasEquipment,
    isExerciseHasEquipmentForWeight,
    isWeightsRequired,
    isUserHasEquipment,
    isUserHasEquipmentForWeight,
  } = getUserEquipmentParams(exercise, user);

  if (!isExerciseHasEquipment && !isExerciseHasEquipmentForWeight) {
    result = true;
  } else if (isUserHasEquipment && !isWeightsRequired) {
    result = true;
  } else if (!isExerciseHasEquipment && isExerciseHasEquipmentForWeight && isUserHasEquipmentForWeight) {
    result = true;
  } else if (!isExerciseHasEquipment && !isWeightsRequired) {
    result = true;
  } else if (
    isExerciseHasEquipment &&
    isExerciseHasEquipmentForWeight &&
    isUserHasEquipment &&
    isUserHasEquipmentForWeight
  ) {
    result = true;
  }

  return result;
}

export function getAverageDuration(exercise: IExerciseStatistics, type: 'set' | 'repeat') {
  return type === 'set'
    ? `${((exercise.averageDuration * exercise.repeats) / exercise.sets || 0).toFixed(1)}`
    : `${exercise.averageDuration.toFixed(1)}`;
}

export function getAverageRepeatsInSet(repeats?: number, sets?: number) {
  return repeats && sets ? Math.round(repeats / sets) : 0;
}

export function isPrevExerciseSame(exercises: IExerciseDone[], index: number, id?: string) {
  return id && exercises[index - 1] ? id === exercises[index - 1].exercise?._id : false;
}

export function isSetCreatable(choosenExercises: IExerciseChoosen[], index: number, id?: string) {
  return !isPrevExerciseSame(choosenExercises, index, id) && index > 0 && index + 1 === choosenExercises.length;
}

export function generateTimeline(exercises: IExerciseDone[], start: Date | string | null, ratio: number) {
  const allSteps: ITimelineStep[] = [];

  exercises.forEach((exercise, index) => {
    if (!exercise.dateUpdated || !exercise.duration || !start) return;

    allSteps.push({
      left:
        index === 0
          ? 0
          : (new Date(exercise.dateUpdated).getTime() - exercise.duration * 1000 - new Date(start).getTime()) / ratio,
      right:
        index === exercises.length - 1 && exercises.length !== 1
          ? 0
          : (new Date(exercise.dateUpdated).getTime() - new Date(start).getTime()) / ratio,
      type: 'exercise',
    });
  });

  allSteps.forEach((step, index) => {
    if (!allSteps[index - 1]?.right) return;

    allSteps.push({ left: allSteps[index - 1].right, right: step.left, type: 'rest' });
  });

  return allSteps.sort((a, b) => a.left - b.left);
}

export function filterExercisesByTitleAndMuscle(exercises: IExercise[], title: string, muscle?: string, user?: IUser) {
  return exercises.filter((exercise) => {
    const isEquipmentMatches = isUserEquipmentMatches(exercise, user);

    const titleFilter =
      (exercise.title.toLowerCase().includes(title.toLocaleLowerCase()) ||
        exercise.title_en?.toLowerCase().includes(title.toLocaleLowerCase())) &&
      isEquipmentMatches;
    const muscleFilter = exercise.muscles?.some((group) => group._id === muscle) && isEquipmentMatches;

    return muscle ? muscleFilter && titleFilter : titleFilter;
  });
}

export function getAvailableExerciseWeights(exercise: IExercise, user: IUser) {
  const equipments = user.equipments?.filter((equipment) =>
    exercise.equipmentForWeight?.some((eq) => eq._id === equipment.equipment?._id)
  );

  if (!equipments?.length) return [];

  const weights = getWeightsForUserEquipment(equipments);

  return exercise.isWeightsRequired ? weights : [0, ...weights];
}

export function getExercisesToChooseDefaultWeight(exercises: IExercise[], userEquipments?: IUserEquipment[]) {
  const equipmentsForWeight = userEquipments?.filter((eq) => eq.weights?.length);
  const exercisesWithWeight = exercises.filter((ex) => ex.isWeights);

  const tableData = exercisesWithWeight.map((ex) => {
    const exerciseEquipment = ex.equipmentForWeight
      ?.filter((eq) => equipmentsForWeight?.some((equipment) => equipment.equipment?._id === eq._id))
      .map((exToMap) => exToMap._id);

    const availableEquipment = equipmentsForWeight?.filter((eq) =>
      exerciseEquipment?.some((equipmentId) => equipmentId === eq.equipment?._id)
    );

    const options = getWeightsForUserEquipment(availableEquipment);

    return {
      _id: ex._id,
      title: ex.title,
      title_en: ex.title_en,
      options: ex.isWeightsRequired ? options : [0, ...options],
    };
  });

  return tableData.filter((ex) => ex.options.length > 1);
}

export function getDefaultExerciseWeight(exercise: IExercise, user: IUser, weights?: number[]) {
  let weight = weights?.length && exercise.isWeightsRequired ? weights[0] : 0;

  if (user.defaultWeights && exercise._id && Object.hasOwn(user.defaultWeights, exercise._id)) {
    weight = user.defaultWeights[exercise._id];
  }

  return weight;
}

export function addSetToExercises(exercises: IExerciseChoosen[]) {
  const set = exercises.slice(-2).map((exercise) => {
    return { ...exercise, _id: createTempId() };
  });

  return [...exercises, ...set];
}

export function updateExercisesIndex(exercises: IExerciseChoosen[], index: number) {
  if (index <= 0 || index >= exercises.length) return exercises;

  const updatedExercises = [...exercises];
  const exerciseToMove = updatedExercises[index];

  updatedExercises.splice(index, 1);
  updatedExercises.splice(index - 1, 0, exerciseToMove);

  return updatedExercises;
}

export function updateExerciseField<T extends keyof IExerciseDone>(
  exercises: IExerciseChoosen[],
  field: T,
  value: IExerciseDone[T],
  id?: string
): IExerciseChoosen[] {
  return exercises.map((exercise) => (exercise._id === id ? { ...exercise, [field]: value } : exercise));
}

export function generateExerciseChoosen(repeats: number, weight?: number, exercise?: IExercise): IExerciseChoosen {
  return {
    _id: createTempId(),
    exercise: {
      _id: exercise?._id || '',
      title: exercise?.title || '',
      title_en: exercise?.title_en || '',
      muscles: exercise?.muscles || [],
      isWeights: exercise?.isWeights || false,
      isWeightsRequired: exercise?.isWeightsRequired || false,
      equipmentForWeight: exercise?.equipmentForWeight,
    },
    repeats,
    weight,
  };
}
