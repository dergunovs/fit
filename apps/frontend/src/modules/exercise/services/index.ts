import { ComputedRef, Ref } from 'vue';
import {
  API_EXERCISE,
  API_EXERCISE_ALL,
  API_EXERCISE_CUSTOM,
  TGetExercisesDTO,
  TGetExercisesCustomDTO,
  TGetExercisesQueryDTO,
  TGetExerciseDTO,
  TPostExerciseDTO,
  TPostExerciseDataDTO,
  TUpdateExerciseDTO,
  TUpdateExerciseDataDTO,
  TDeleteExerciseDTO,
} from 'fitness-tracker-contracts';
import { useMutation, useQuery, api } from 'mhz-helpers';

export const exerciseService = {
  getMany: (page: Ref<number>) => {
    return useQuery({
      queryKey: [API_EXERCISE, page],
      queryFn: async () => {
        const params: TGetExercisesQueryDTO = { page: page.value };

        const { data } = await api.get<TGetExercisesDTO>(API_EXERCISE, { params });

        return data;
      },
    });
  },

  getAll: () => {
    return useQuery({
      queryKey: [API_EXERCISE, API_EXERCISE_ALL],
      queryFn: async () => {
        const { data } = await api.get<TGetExercisesDTO>(API_EXERCISE_ALL);

        return data.data;
      },
    });
  },

  getCustom: () => {
    return useQuery({
      queryKey: [API_EXERCISE, API_EXERCISE_CUSTOM],
      queryFn: async () => {
        const { data } = await api.get<TGetExercisesCustomDTO>(API_EXERCISE_CUSTOM);

        return data.data;
      },
    });
  },

  getOne: (options: object, id?: ComputedRef<string | undefined>) => {
    return useQuery({
      queryKey: [API_EXERCISE, id],
      queryFn: async () => {
        if (!id?.value) return null;

        const { data } = await api.get<TGetExerciseDTO>(`${API_EXERCISE}/${id.value}`);

        return data.data;
      },
      ...options,
    });
  },

  create: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (formData: TPostExerciseDataDTO) => {
        const { data } = await api.post<TPostExerciseDTO>(API_EXERCISE, formData);

        return data;
      },
      ...options,
    });
  },

  update: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (formData: TUpdateExerciseDataDTO) => {
        const { data } = await api.patch<TUpdateExerciseDTO>(`${API_EXERCISE}/${formData._id}`, formData);

        return data;
      },
      ...options,
    });
  },

  delete: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (id: string) => {
        const { data } = await api.delete<TDeleteExerciseDTO>(`${API_EXERCISE}/${id}`);

        return data;
      },
      ...options,
    });
  },
};
