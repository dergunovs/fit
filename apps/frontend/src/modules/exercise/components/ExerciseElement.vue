<template>
  <div
    v-show="!props.isPassing || (props.isPassing && !props.isDone)"
    :class="$style.exercise"
    :data-hide="props.isHideTitle"
    :data-edit="props.isEdit"
  >
    <ExerciseMuscleColors
      v-if="props.exercise.exercise?.muscles && !props.isHideTitle"
      :muscles="props.exercise.exercise.muscles"
      data-test="exercise-muscle-colors"
    />

    <UiFlex column :gap="props.isEdit ? '4' : '2'" align="flex-start" grow>
      <ExerciseElementTitle
        v-if="!props.isHideTitle"
        :exercise="props.exercise.exercise"
        :exercisesCount="props.exercisesCount"
        :index="props.index"
        :isEdit="props.isEdit"
        :isPassing="props.isPassing"
        :isCurrent="props.isCurrent"
        data-test="exercise-element-title"
      />

      <ExerciseButtons
        v-if="props.exercise._id"
        :isEdit="props.isEdit"
        :isPassing="props.isPassing"
        :repeats="props.exercise.repeats"
        :weight="props.exercise.weight"
        :index="props.index"
        :isLast="props.isLast"
        :isSetCreatable="props.isSetCreatable"
        :isDone="props.isDone"
        :isToFailure="props.isToFailure"
        :duration="props.duration"
        :isFutureActivity="props.isFutureActivity"
        :isWeights="props.exercise.exercise?.isWeights"
        :isCurrent="props.isCurrent"
        :isActive="props.isActive"
        :isOnlyOne="props.exercisesCount === 1 && props.exercisesCount === index + 1"
        @createSet="emit('createSet')"
        @delete="emit('delete', props.exercise._id)"
        @setIndex="(updatedIndex) => emit('setIndex', updatedIndex)"
        @editRepeats="toggleExerciseRepeats"
        @editWeight="toggleExerciseWeight"
        @setIsToFailure="(isToFailure) => emit('setIsToFailure', isToFailure)"
        data-test="exercise-buttons"
      />

      <Transition name="slide" mode="out-in">
        <ExerciseRepeats
          v-show="isShowEditRepeats"
          :modelValue="props.exercise.repeats"
          :baseRepeat="props.exercise.repeats"
          isTall
          @update:modelValue="updateRepeats"
          data-test="exercise-repeats"
        />
      </Transition>

      <Transition name="slide" mode="out-in">
        <ExerciseWeight
          v-if="props.exercise.exercise && user"
          v-show="isShowEditWeight"
          @update:modelValue="updateWeight"
          :modelValue="props.exercise.weight"
          :options="getAvailableExerciseWeights(props.exercise.exercise, user)"
          data-test="exercise-weight"
        />
      </Transition>

      <slot></slot>
    </UiFlex>
  </div>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import { IExerciseChoosen } from 'fitness-tracker-contracts';
import { UiFlex } from 'mhz-ui';

import ExerciseElementTitle from '@/exercise/components/ExerciseElementTitle.vue';
import ExerciseMuscleColors from '@/exercise/components/ExerciseMuscleColors.vue';
import ExerciseRepeats from '@/exercise/components/ExerciseRepeats.vue';
import ExerciseWeight from '@/exercise/components/ExerciseWeight.vue';
import ExerciseButtons from '@/exercise/components/ExerciseButtons.vue';

import { getAvailableExerciseWeights } from '@/exercise/helpers';
import { useAuthCheck } from '@/auth/composables';

interface IProps {
  exercise: IExerciseChoosen;
  index: number;
  isEdit?: boolean;
  isPassing?: boolean;
  isSetCreatable?: boolean;
  isLast?: boolean;
  isHideTitle?: boolean;
  isDone?: boolean;
  isToFailure?: boolean;
  isActive?: boolean;
  isCurrent?: boolean;
  duration?: number;
  isFutureActivity?: boolean;
  exercisesCount?: number;
}

interface IEmit {
  delete: [id: string];
  createSet: [];
  setIndex: [index: number];
  setRepeats: [repeats: number];
  setWeight: [weight: number];
  setIsToFailure: [isToFailure: boolean];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();

const { user } = useAuthCheck();

const isShowEditRepeats = ref(false);
const isShowEditWeight = ref(false);

function toggleExerciseRepeats() {
  isShowEditRepeats.value = !isShowEditRepeats.value;
  isShowEditWeight.value = false;
}

function toggleExerciseWeight() {
  isShowEditWeight.value = !isShowEditWeight.value;
  isShowEditRepeats.value = false;
}

function updateRepeats(repeats: number) {
  emit('setRepeats', repeats);

  toggleExerciseRepeats();
}

function updateWeight(weight: number) {
  emit('setWeight', weight);

  toggleExerciseWeight();
}
</script>

<style module lang="scss">
.exercise {
  position: relative;
  display: flex;
  gap: 12px;
  justify-content: space-between;
  padding: 6px 12px;
  background-color: var(--color-gray-light-extra);
  border-bottom: 1px solid var(--color-gray);
  border-radius: 8px;

  &[data-edit='true'] {
    padding: 8px 12px;
  }

  &[data-hide='true'] {
    margin-top: -14px;
    border-top-left-radius: 0;
    border-top-right-radius: 0;
  }
}

:global(.dark) {
  .exercise {
    background-color: var(--color-black);
    border: 1px solid var(--color-gray-dark-extra);

    &[data-hide='true'] {
      border-top: none;
    }
  }
}
</style>

<style lang="scss">
.slide-enter-active,
.slide-leave-active {
  transition: all 200ms ease;
}

.slide-enter-from,
.slide-leave-to {
  opacity: 0;
  transform: translateY(-10px);
}
</style>
