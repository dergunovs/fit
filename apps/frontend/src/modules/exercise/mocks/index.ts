import { vi } from 'vitest';
import {
  TDeleteExerciseDTO,
  TGetExercisesDTO,
  TPostExerciseDataDTO,
  TPostExerciseDTO,
  TUpdateExerciseDataDTO,
  TUpdateExerciseDTO,
} from 'fitness-tracker-contracts';

import { exerciseService } from '@/exercise/services';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { EXERCISE_FIXTURE, EXERCISE_FIXTURE_CUSTOM, EXERCISES_FIXTURE } from '@/exercise/fixtures';
import { IOnSuccess } from '@/common/interface';

const spyGetExercise = vi.spyOn(exerciseService, 'getOne').mockReturnValue(mockQueryReply(EXERCISE_FIXTURE));

const mockGetExercisesData: TGetExercisesDTO = { data: EXERCISES_FIXTURE, total: EXERCISES_FIXTURE.length };

const spyGetExercises = vi
  .spyOn(exerciseService, 'getMany')
  .mockImplementation(() => mockQueryReply(mockGetExercisesData));

const spyGetExercisesAll = vi
  .spyOn(exerciseService, 'getAll')
  .mockImplementation(() => mockQueryReply(EXERCISES_FIXTURE));

const spyGetExercisesCustom = vi
  .spyOn(exerciseService, 'getCustom')
  .mockImplementation(() => mockQueryReply([EXERCISE_FIXTURE_CUSTOM]));

const spyCreateExercise = vi.fn();
const spyUpdateExercise = vi.fn();
const spyDeleteExercise = vi.fn();

const mockOnSuccess: IOnSuccess = {
  create: undefined,
  update: undefined,
  delete: undefined,
};

vi.spyOn(exerciseService, 'create').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.create = options.onSuccess;

  return mockMutationReply<TPostExerciseDTO, TPostExerciseDataDTO>(spyCreateExercise);
});

vi.spyOn(exerciseService, 'update').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.update = options.onSuccess;

  return mockMutationReply<TUpdateExerciseDTO, TUpdateExerciseDataDTO>(spyUpdateExercise);
});

vi.spyOn(exerciseService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteExerciseDTO, string>(spyDeleteExercise);
});

export {
  spyGetExercise,
  spyGetExercises,
  spyGetExercisesCustom,
  spyGetExercisesAll,
  spyCreateExercise,
  spyUpdateExercise,
  spyDeleteExercise,
  mockOnSuccess,
  mockGetExercisesData,
};
