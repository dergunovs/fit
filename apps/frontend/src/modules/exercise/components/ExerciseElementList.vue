<template>
  <div>
    <UiFlex column>
      <TransitionGroup name="list" tag="div" :class="$style.block">
        <ExerciseElement
          v-for="(exercise, index) in props.exercises as IExerciseDone[]"
          :key="exercise._id"
          :exercise="exercise"
          :exercisesCount="props.exercises.length"
          :index="index"
          :isLast="index === props.exercises.length - 1"
          :isSetCreatable="isSetCreatable(props.exercises, index, exercise.exercise?._id)"
          :isEdit="props.isEdit"
          :isDone="exercise.isDone"
          :isPassing="props.isPassing"
          :duration="exercise.duration"
          :isFutureActivity="props.isFutureActivity"
          :isToFailure="exercise.isToFailure"
          :isHideTitle="
            isPrevExerciseSame(props.exercises, index, exercise.exercise?._id) && !props.isEdit && !props.isPassing
          "
          @delete="(id) => emit('delete', id)"
          @createSet="emit('createSet')"
          @setIndex="(updatedIndex) => emit('setIndex', updatedIndex)"
          @setRepeats="(repeats) => emit('setRepeats', repeats, exercise._id)"
          @setWeight="(weight) => emit('setWeight', weight, exercise._id)"
          data-test="exercise-element"
        >
          <slot :exercise="exercise" :index="index"></slot>
        </ExerciseElement>
      </TransitionGroup>
    </UiFlex>
  </div>
</template>

<script setup lang="ts">
import { UiFlex } from 'mhz-ui';
import { IExerciseChoosen, IExerciseDone } from 'fitness-tracker-contracts';

import ExerciseElement from '@/exercise/components/ExerciseElement.vue';

import { isSetCreatable, isPrevExerciseSame } from '@/exercise/helpers';

interface IProps {
  exercises: IExerciseChoosen[] | IExerciseDone[];
  isEdit?: boolean;
  isPassing?: boolean;
  isFutureActivity?: boolean;
}

interface IEmit {
  delete: [id: string];
  createSet: [];
  setIndex: [index: number];
  setRepeats: [repeats: number, id?: string];
  setWeight: [weight: number, id?: string];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();
</script>

<style module lang="scss">
.block {
  display: flex;
  flex-direction: column;
  gap: 8px;
}
</style>

<style lang="scss">
.list-enter-active,
.list-leave-active,
.list-move {
  transition: all 300ms ease;
}

.list-enter-from,
.list-leave-to {
  opacity: 0;
  transform: translateY(-20px);
}

.list-leave-active {
  position: absolute;
}
</style>
