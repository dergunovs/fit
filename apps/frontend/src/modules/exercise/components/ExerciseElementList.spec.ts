import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { dataTest } from 'mhz-helpers';

import ExerciseElementList from './ExerciseElementList.vue';
import ExerciseElement from './ExerciseElement.vue';

import { wrapperFactory } from '@/common/test';
import { EXERCISES_CHOOSEN_FIXTURE } from '@/exercise/fixtures';

const exercise = dataTest('exercise-element');

let wrapper: VueWrapper<InstanceType<typeof ExerciseElementList>>;

beforeEach(() => {
  wrapper = wrapperFactory(ExerciseElementList, { exercises: EXERCISES_CHOOSEN_FIXTURE });
});

enableAutoUnmount(afterEach);

describe('ExerciseElementList', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ExerciseElementList)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shows choosen exercises', async () => {
    expect(wrapper.findAll(exercise).length).toBe(EXERCISES_CHOOSEN_FIXTURE.length);
    expect(wrapper.findComponent<typeof ExerciseElement>(exercise).props('exercise')).toStrictEqual(
      EXERCISES_CHOOSEN_FIXTURE[0]
    );
  });

  it('emits choosen exercise events', async () => {
    const INDEX = 1;
    const ID = EXERCISES_CHOOSEN_FIXTURE[0]._id;
    const REPEATS = 8;
    const WEIGHT = 20;

    expect(wrapper.emitted()).not.toHaveProperty('createSet');
    expect(wrapper.emitted()).not.toHaveProperty('delete');
    expect(wrapper.emitted()).not.toHaveProperty('setIndex');
    expect(wrapper.emitted()).not.toHaveProperty('setRepeats');
    expect(wrapper.emitted()).not.toHaveProperty('setWeight');

    wrapper.findComponent<typeof ExerciseElement>(exercise).vm.$emit('createSet');

    expect(wrapper.emitted('createSet')).toHaveLength(1);
    expect(wrapper.emitted()['createSet'][0]).toStrictEqual([]);

    wrapper.findComponent<typeof ExerciseElement>(exercise).vm.$emit('delete', ID);

    expect(wrapper.emitted('delete')).toHaveLength(1);
    expect(wrapper.emitted()['delete'][0]).toStrictEqual([ID]);

    wrapper.findComponent<typeof ExerciseElement>(exercise).vm.$emit('setIndex', INDEX);

    expect(wrapper.emitted('setIndex')).toHaveLength(1);
    expect(wrapper.emitted()['setIndex'][0]).toStrictEqual([INDEX]);

    wrapper.findComponent<typeof ExerciseElement>(exercise).vm.$emit('setRepeats', REPEATS, ID);

    expect(wrapper.emitted('setRepeats')).toHaveLength(1);
    expect(wrapper.emitted()['setRepeats'][0]).toStrictEqual([REPEATS, ID]);

    wrapper.findComponent<typeof ExerciseElement>(exercise).vm.$emit('setWeight', WEIGHT, ID);

    expect(wrapper.emitted('setWeight')).toHaveLength(1);
    expect(wrapper.emitted()['setWeight'][0]).toStrictEqual([WEIGHT, ID]);
  });
});
