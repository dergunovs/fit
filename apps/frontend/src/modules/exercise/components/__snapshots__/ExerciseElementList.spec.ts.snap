// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExerciseElementList > matches snapshot 1`] = `
"<div>
  <div column="">
    <transition-group-stub name="list" tag="div" appear="false" persisted="false" css="true" class="_block_72460c">
      <exercise-element-stub exercise="[object Object]" exercisescount="2" index="0" isedit="false" ispassing="false" issetcreatable="false" islast="false" ishidetitle="false" isfutureactivity="false" data-test="exercise-element"></exercise-element-stub>
      <exercise-element-stub exercise="[object Object]" exercisescount="2" index="1" isedit="false" ispassing="false" issetcreatable="true" islast="true" ishidetitle="false" isfutureactivity="false" data-test="exercise-element"></exercise-element-stub>
    </transition-group-stub>
  </div>
</div>"
`;
